### PR TITLE
QR code scanner (QML component)

### PR DESCRIPTION
**Pull request by igor-sirotin**
Required for https://github.com/status-im/status-desktop/issues/5837

Overrides:
- https://github.com/status-im/status-desktop/pull/5756
- https://github.com/status-im/dotherside/pull/65

### What does the PR do

This is a QR code scanner. Based on [QZXing](https://github.com/ftylitak/qzxing).

Notes:
1. I have an issue on MacOS, that camera loads quite long, about 5-10 seconds. 
Didn't fix it yet, will try to improve it later.
2. I implemented a camera selector, but it didn't fully work for me.
Will try to fix it later.
3. I will also create a PR with changes made for QZXing library. 
Currently pointing to status-im fork.

### Affected areas

StatusQ

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [x] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [x] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

https://user-images.githubusercontent.com/25482501/217012392-c56fd974-0f9e-4407-8c8b-1ff6ce2d4946.mov

### CODE REVIEW

**Comment by igor-sirotin** [comment]
@0x-r4bbit, could you please test this un Windows?
@osmaczko , could you please test this un Linux?

**Comment by 0x-r4bbit** [comment]
@igor-sirotin unfortunately, I don't have a window machine

File: ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    QtObject {
+        id: d
+
+        readonly property var lastSyncDate: new Date(model.timestamp)
+        readonly property int millisecondsFromSync: lastSyncDate - Date.now()
+        readonly property int secondsFromSync: millisecondsFromSync / 1000
+        readonly property int minutesFromSync: secondsFromSync / 60
+        readonly property int daysFromSync: new Date().getDay() - lastSyncDate.getDay()
+
+        readonly property var daysOfWeek: [
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml:33]: https://doc.qt.io/qt-5/qml-qtqml-locale.html#dayName-method

File: ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    subTitle: {
+        if (model.isCurrentDevice)
+            return qsTr("This device");
+
+        if (d.secondsFromSync <= 120)
+            return qsTr("Online now");
+
+        if (d.minutesFromSync <= 60)
+            return qsTr("Online %1 minutes ago").arg(d.minutesFromSync);
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml:54]: Plural:
```suggestion
            return qsTr("Online %n minute(s) ago", "", d.minutesFromSync);
```

File: ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        if (d.daysFromSync == 0)
+            return qsTr("Last seen earlier today");
+
+        if (d.daysFromSync == 1)
+            return qsTr("Last online yesterday");
+
+        if (d.daysFromSync <= 6)
+            return qsTr("Last online [%1]").arg(d.daysOfWeek[d.lastSyncDate.getDay()]);
+
+        return qsTr("Last online %1").arg(d.lastSyncDate.toLocaleDateString(Qt.locale()))
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml:65]: StatusQCore.LocaleUtils.formatFoo()

File: ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    asset.name: !!root.emojiId ? "" : "desktop"
+    asset.emoji: root.emojiId
+    asset.bgColor: root.deviceColor
+    asset.isLetterIdenticon: false
+    // label: qsTr("Next back up in %1 hours")
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/controls/SyncDeviceDelegate.qml:72]: commented out code? (if enabled, needs plural too as above :)

File: ui/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+        onTriggered: {
+            counter = ++counter % 3;
+            if (counter == 0)
+                d.syncingSuccess()
+            else
+                d.syncingFailed()
+        }
+    }
+
+    DSM.StateMachine {
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:126]: Nice :)

File: ui/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+            default:
+                console.error("SetupSyncingPopup: invalid mode %1").arg(root.mode)
+                return undefined;
+            }
+        }
+
+//        DSM.SignalTransition {
+//            targetState: syncingState
+//            signal: d.syncingStarted
+//        }
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:148]: commented out code?

File: ui/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+        StatusButton {
+            id: nextButton
+            visible: !!text
+            text: {
+                // TODO: This button feels wierd.
+                //      I feel like I should press it to continue, while I just need to wait.
+//              //      And the button will close the popup and stop the workflow
+//                i f (displaySyncCodeState.active)
+//                    return qsTr("Close");
+                if (displaySyncCodeState.active)    // TODO: This is temporary, remove
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:309]: Still TODO?

File: ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        property int deviceColorIndex: 0
+    }
+
+    Connections {
+        enabled: d.emojiPopupOpened
+        target: root.emojiPopup
+
+        onEmojiSelected: function (emojiText, atCursor) {
+//            contentItem.channelName.input.asset.isLetterIdenticon = false;
+//            scrollView.channelName.input.asset.emoji = emojiText
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml:45]: commented out code?

File: ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    onOpened: {
+        colorGrid.selectedColorIndex = d.deviceColorIndex
+        nameInput.text = deviceModel.name
+    }
+
+    rightButtons: [
+        StatusButton {
+            text: "Done"
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml:61]: qsTr()

File: ui/app/AppLayouts/Profile/popups/setupsyncing/EnterSyncCode.qml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+            readonly property bool attemptsFlag: root.pairingFailsCount < 2
+
+            visible: root.pairingFailed
+            showThinkingEmoji: attemptsFlag
+            primaryText: attemptsFlag ? qsTr("Hmm, that didn't work")
+                                      : qsTr("That still didn’t work")
+            secondaryText: root.pairingErrorMessage
+//            secondaryText: attemptsFlag ? qsTr("Please try pasting the sync code again.")
+//                                        : qsTr("Double check to make sure the code isn’t expired.")
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/setupsyncing/EnterSyncCode.qml:52]: commented out code?

File: ui/app/AppLayouts/Profile/popups/setupsyncing/ScanSyncCode.qml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        Layout.fillHeight: true
+
+        horizontalAlignment: Text.AlignHCenter
+        verticalAlignment: Text.AlignVCenter
+
+        linkColor: Theme.palette.primaryColor1
+        onLinkActivated: Global.openLink(link)
+
+        text: {
+            const link = '<a href="https://github.com/status-im/status-desktop/pull/5756">#5756 Scan QR code</a>';
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/setupsyncing/ScanSyncCode.qml:23]: "Scan QR code" needs qsTr()

File: ui/app/AppLayouts/Profile/popups/setupsyncing/Syncing.qml [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        model: root.devicesStore.devicesModel
+        spacing: 4
+        clip: true
+
+        delegate: SyncDeviceDelegate {
+            width: ListView.view.width
+            enabled: false
+        }
+    }
+
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/popups/setupsyncing/Syncing.qml:122]: empty line

File: ui/app/AppLayouts/Profile/views/SyncingView.qml [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+
+        Connections {
+            target: devicesStore.devicesModule
+
+            function onUserAuthenticated(pin, password, keyUid) {
+
+                // WARNING: Test keycard support
+                console.log("<<< user ath", keyUid, pin, password)
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/views/SyncingView.qml:88]: Nope, don't show password to console :)

File: ui/app/AppLayouts/Profile/views/SyncingView.qml [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        ListView {
+            id: listView
+            Layout.fillWidth: true
+            // Layout.maximumHeight: 400
+            Layout.topMargin: 17
+            Layout.bottomMargin: 17
+
+            implicitHeight: contentHeight
+            // contentWidth: contentItem.childrenRect.width
+            // contentHeight: contentItem.childrenRect.height
```
Comment (caybro) [inline ui/app/AppLayouts/Profile/views/SyncingView.qml:120]: commented out code

File: ui/nim-status-client.pro [hunk 15]
```diff
@@ -45,3 +45,6 @@ else: unix:!android: target.path = /opt/$${TARGET}/bin
 !isEmpty(target.path): INSTALLS += target
 
 RESOURCES += resources.qrc
+
+DISTFILES += \
+    app/AppLayouts/Profile/popups/setupsyncing/ErrorMessage.qml
```
Comment (caybro) [inline ui/nim-status-client.pro:50]: Not needed

File: src/app_service/service/devices/service.nim [hunk 16]
```diff
@@ -97,3 +107,26 @@ QtObject:
...
+
+  proc inputConnectionStringForBootstrapping*(self: Service, connectionString: string): string =
+    let configJSON = %* {
+      "keyUid": singletonInstance.userProfile.getKeyUid(),
+      "keystorePath": main_constants.ROOTKEYSTOREDIR,
+      "password": string_utils.hashString("qwercqwerc")
+    }
+    echo "<<< inputConnectionStringForBootstrapping call " & $configJSON & " " & connectionString
+    let result = status_go.inputConnectionStringForBootstrapping(connectionString, $configJSON) 
+    echo "<<< inputConnectionStringForBootstrapping result " & $result
```
Comment (caybro) [inline src/app_service/service/devices/service.nim:130]: debug leftovers?

**Review by caybro — CHANGES_REQUESTED** [review]
Awesome amount of work, some minor things inline

**Comment by igor-sirotin** [comment]
@caybro, it seems that I pushed my changes from other branch to this PR.
I'm really sorry for this, I've removed redundant commits.
But I will use your comments and fix them in my code anyway 🙂

File: ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+    Camera {
+        id: camera
+        captureMode: Camera.CaptureVideo
+        focus {
+            focusMode: CameraFocus.FocusContinuous
+            focusPointMode: CameraFocus.FocusPointAuto
+        }
+        onDeviceIdChanged: {
+            console.log("<<< deviceId =", deviceId)
```
Comment (caybro) [inline ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml:46]: debug?

File: ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+    //    }
+
+
+    StatusBaseText {
+        anchors.fill: parent
+        horizontalAlignment: Text.AlignHCenter
+        verticalAlignment: Text.AlignVCenter
+        color: Theme.palette.dangerColor1
+        visible: camera.availability !== Camera.Available
+        text: qsTr("Camera unavailable: ") + camera.errorString
```
Comment (caybro) [inline ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml:166]:
```suggestion
        text: qsTr("Camera unavailable: %1").arg(camera.errorString)
```

**Review by caybro — COMMENTED** [review]
Looks better :)

**Comment by igor-sirotin** [comment]
@caybro, thanks for comments, fixed that!

**Review by caybro — APPROVED** [review]
Cool!

**Comment by osmaczko** [comment]
@osmaczko , could you please test this un Linux?

It works! Although, it took a while to find the right angle :sweat_smile: 

https://user-images.githubusercontent.com/33099791/217190291-29b125e4-db8f-44fc-a707-a0e3696ddaf4.mp4

**Review by osmaczko — COMMENTED** [review]
The import check is broken because StatusQ knows nothing about DoOtherSide. This needs to be addressed, following https://github.com/status-im/status-desktop/issues/9125 rationale.

**Review by alaibe — APPROVED** [review]

**Review by caybro — COMMENTED** [review]
A big question: do we need this in DOtherSide? A better place would imho be `status-desktop/ui/StatusQ/src`, now that we decoupled those

cc @micieslak

**Comment by richard-ramos** [comment]
Very cool PR! can confirm it works on linux!
Something that we should think in adding in the future is a webcam selector, as it currently does not have a way to choose which cam to use for scanning the QR code.

**Comment by micieslak** [comment]
Great job! It works for me from Sandbox, but unfortunately there are still problems to run the main app:
```

/usr/bin/ld: vendor/DOtherSide/build/lib/libDOtherSideStatic.a(DOtherSide.cpp.o): in function `dos_qguiapplication_create':
DOtherSide.cpp:(.text+0x7cc9): undefined reference to `QZXing::registerQMLTypes()'
collect2: error: ld returned 1 exit status
Error: execution of an external program failed: 'g++  @nim_status_client_linkerArgs.txt'
make: *** [Makefile:375: bin/nim_status_client] Error 1
01:00:04: The process "/usr/bin/make" exited with code 2.
```
I was playing a bit to solve that but so far no results. Some nim experts are needed here, I suppose...

**Comment by micieslak** [comment]
> A big question: do we need this in DOtherSide? A better place would imho be `status-desktop/ui/StatusQ/src`, now that we decoupled those
> 
> cc @micieslak

If we want to use it in the main app (and we do, I suppose), we need to compile it via DOtherSide for the main app. At least I can't see any better solution for now.

**Comment by caybro** [comment]
Definitely needs a rebase

**Comment by richard-ramos** [comment]
This should build in Linux now!

**Comment by caybro** [comment]
> This should build in Linux now!

Nope...
```
[2023-03-06T13:59:07.606Z] ERROR: ldd outputLine: "libodbc.so.2 => not found"

[2023-03-06T13:59:07.606Z] ERROR: for binary: "/opt/qt/5.15.2/gcc_64/plugins/sqldrivers/libqsqlodbc.so"

[2023-03-06T13:59:07.606Z] ERROR: Please ensure that all libraries can be found by ldd. Aborting.
```

I don't know what on earth needs a Qt ODBC driver but it's not found

**Comment by richard-ramos** [comment]
Interestingly enough, if I execute `ldd` in `./bin/nim_status_client`, I get the following list of libraries:
```
        linux-vdso.so.1 (0x00007ffdb69eb000)
        libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3 (0x00007fb91ec54000)
        libssl.so.3 => /lib/x86_64-linux-gnu/libssl.so.3 (0x00007fb91ebb0000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007fb91eb3a000)
        libQt5QuickControls2.so.5 => not found
        libQt5Quick.so.5 => /lib/x86_64-linux-gnu/libQt5Quick.so.5 (0x00007fb91e5c0000)
        libQt5Qml.so.5 => /lib/x86_64-linux-gnu/libQt5Qml.so.5 (0x00007fb91e134000)
        libQt5Multimedia.so.5 => not found
        libQt5Network.so.5 => /lib/x86_64-linux-gnu/libQt5Network.so.5 (0x00007fb91df83000)
        libQt5Gui.so.5 => /lib/x86_64-linux-gnu/libQt5Gui.so.5 (0x00007fb91d8a5000)
        libQt5Core.so.5 => /lib/x86_64-linux-gnu/libQt5Core.so.5 (0x00007fb91d346000)
        libstatus.so.0 => not found
        libkeycard.so => not found
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fb91d25d000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fb91d033000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fb91d013000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb91cdeb000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb9208a3000)
        libQt5QmlModels.so.5 => /lib/x86_64-linux-gnu/libQt5QmlModels.so.5 (0x00007fb91cd60000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fb91cd42000)
        libgssapi_krb5.so.2 => /lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007fb91ccee000)
        libGL.so.1 => /lib/x86_64-linux-gnu/libGL.so.1 (0x00007fb91cc67000)
        libpng16.so.16 => /lib/x86_64-linux-gnu/libpng16.so.16 (0x00007fb91cc2c000)
        libharfbuzz.so.0 => /lib/x86_64-linux-gnu/libharfbuzz.so.0 (0x00007fb91cb5d000)
        libmd4c.so.0 => /lib/x86_64-linux-gnu/libmd4c.so.0 (0x00007fb91cb4b000)
        libdouble-conversion.so.3 => /lib/x86_64-linux-gnu/libdouble-conversion.so.3 (0x00007fb91cb34000)
        libicui18n.so.70 => /lib/x86_64-linux-gnu/libicui18n.so.70 (0x00007fb91c805000)
        libicuuc.so.70 => /lib/x86_64-linux-gnu/libicuuc.so.70 (0x00007fb91c60a000)
        libpcre2-16.so.0 => /lib/x86_64-linux-gnu/libpcre2-16.so.0 (0x00007fb91c580000)
        libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007fb91c4b1000)
        libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007fb91c377000)
        libkrb5.so.3 => /lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007fb91c2aa000)
        libk5crypto.so.3 => /lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007fb91c27b000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007fb91c275000)
        libkrb5support.so.0 => /lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007fb91c267000)
        libGLdispatch.so.0 => /lib/x86_64-linux-gnu/libGLdispatch.so.0 (0x00007fb91c1af000)
        libGLX.so.0 => /lib/x86_64-linux-gnu/libGLX.so.0 (0x00007fb91c179000)
        libfreetype.so.6 => /lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007fb91c0b1000)
        libgraphite2.so.3 => /lib/x86_64-linux-gnu/libgraphite2.so.3 (0x00007fb91c08a000)
        libicudata.so.70 => /lib/x86_64-linux-gnu/libicudata.so.70 (0x00007fb91a46c000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007fb91a465000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007fb91a44f000)
        libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x00007fb91a30f000)
        libbrotlidec.so.1 => /lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007fb91a301000)
        libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007fb91a2d7000)
        libbrotlicommon.so.1 => /lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007fb91a2b4000)
        libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007fb91a2ac000)
        libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007fb91a2a4000)
        libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007fb91a28c000)
        libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007fb91a27f000)
```

and no mention of libodbc.so

**Comment by caybro** [comment]
> libqsqlodbc.so

But this is not the issue; you want to look at the runtime dependencies of the `libqsqlodbc.so` plugin, and here on my system:

```
ltinkl@zephyrus14:/usr/lib64/qt5/plugins/sqldrivers$ ldd libqsqlodbc.so  | grep -i odbc
	libodbc.so.2 => /lib64/libodbc.so.2 (0x00007f85ec5a4000)
```

**Comment by caybro** [comment]
In other words, this seems more like a packaging problem, where we ship _all_ Qt plugins but some of them (namely the libqsqlodbc.so plugin which we don't even need in our app) have unsatisfied runtime deps

**Comment by richard-ramos** [comment]
Interesting! i tried to build the app without using the lib and i'm running into the same error!
https://github.com/status-im/status-desktop/pull/9753

**Comment by caybro** [comment]
> Interesting! i tried to build the app without using the lib and i'm running into the same error! #9753

Interesting... my fresh new PR builds just fine here: https://github.com/status-im/status-desktop/pull/9666; others don't seem to be having this problem either

**Pull request merged**